api/src/main/kotlin/org/jellyfin/apiclient/api/client/sockets/WebSocketApi.kt

```diff
@@ -0,0 +1,221 @@
+package org.jellyfin.apiclient.api.client.sockets
+
+import io.ktor.client.features.websocket.*
+import io.ktor.client.request.*
+import io.ktor.http.*
+import io.ktor.http.cio.websocket.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.BroadcastChannel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.Channel.Factory.BUFFERED
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.flow.*
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.*
+import kotlinx.serialization.serializer
+import org.jellyfin.apiclient.api.client.KtorClient
+import org.jellyfin.apiclient.model.socket.*
+import org.slf4j.LoggerFactory
+
+/**
+ * Provides realtime communication with the Jellyfin server.
+ * Automatically connects when subscribed to or when messages are sent and disconnects when no subscribers left.
+ *
+ * When the connections drops it will reconnect automatically.
+ * Will not react to access token or server url changes, call .reconnect() to act on these changes.
+ *
+ * The user should verify the access token is correct as the server does not respond to bad authorization.
+ */
+@FlowPreview
+@ExperimentalCoroutinesApi
+@ExperimentalSerializationApi
+class WebSocketApi(
+	private val api: KtorClient
+) {
+	private val logger = LoggerFactory.getLogger("WebSocketApi")
+
+	private val json = Json {
+		// Based on DefaultJson
+		isLenient = false
+		ignoreUnknownKeys = true
+		allowSpecialFloatingPointValues = true
+		useArrayPolymorphism = false
+		encodeDefaults = true
+	}
+
+	private var socketJob: Job? = null
+
+	private val incomingMessageChannel by lazy {
+		BroadcastChannel<IncomingSocketMessage>(BUFFERED)
+	}
+
+	private val outgoingMessageChannel by lazy {
+		Channel<String>(BUFFERED)
+	}
+
+	private suspend fun ReceiveChannel<Frame>.read() = consumeAsFlow()
+		.onEach {
+			val message = it.readSocketMessage() ?: return@onEach
+
+			// Send message
+			incomingMessageChannel.send(message)
+		}
+		.catch { it.printStackTrace() }
+		.onCompletion {
+			// Reconnect
+			// TODO Only reconnect if listeners exist
+			println("END")
+			reconnect()
+		}
+		.collect()
+
+	private suspend fun SendChannel<Frame>.write() = outgoingMessageChannel
+		.receiveAsFlow()
+		.onEach { text ->
+			logger.info("Sending message {}", text)
+			send(Frame.Text(text))
+		}
+		.catch { it.printStackTrace() }
+		.collect()
+
+	private fun ensureConnected() {
+		if (socketJob == null) reconnect()
+	}
+
+	fun reconnect() {
+		// Get base url and access token and check if they're set
+		val baseUrl = checkNotNull(api.baseUrl)
+		val accessToken = checkNotNull(api.accessToken)
+
+		println("RECONNECT")
+
+		// Close existing socket
+		socketJob?.cancel()
+
+		// Open socket
+		socketJob = GlobalScope.launch(Dispatchers.IO) {
```

Comment:
I think I used GlobalScope because the job should not become child job of the current context because it could close the socket. I'm not sure if that's a good argument to using the GlobalScope though.

Reply:
What about adding another explicit, private scope to that class?

Reply:
I think I needed to use the unconfined dispatcher, seems to work with that (see latest commit). I'll need to do some intensive testing to make sure this works without leaks.